File: litho-editor-flipper/src/main/java/com/facebook/litho/editor/flipper/DataUtils.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho.editor.flipper;

import static com.facebook.flipper.plugins.inspector.InspectorValue.Type.Color;

import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import com.facebook.flipper.core.FlipperObject;
import com.facebook.flipper.plugins.inspector.InspectorValue;
import com.facebook.flipper.plugins.inspector.Named;
import com.facebook.litho.SpecGeneratedComponent;
import com.facebook.litho.StateContainer;
import com.facebook.litho.annotations.Prop;
import com.facebook.litho.annotations.ResType;
import com.facebook.litho.annotations.State;
import com.facebook.litho.drawable.ComparableColorDrawable;
import java.lang.reflect.Field;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import javax.annotation.Nullable;

public class DataUtils {

  static List<Named<FlipperObject>> getPropData(Object node) {
    final FlipperObject.Builder props = new FlipperObject.Builder();
    List<Named<FlipperObject>> data = new ArrayList<>();

    boolean hasProps = false;

    boolean isSpecComponent = node instanceof SpecGeneratedComponent;

    for (Field f : node.getClass().getDeclaredFields()) {
      f.setAccessible(true);

      Object prop;
      try {
        prop = f.get(node);
      } catch (IllegalAccessException e) {
        continue;
      }
      String propName = f.getName();

      final Prop annotation = f.getAnnotation(Prop.class);
      if (isSpecComponent && annotation == null) {
        // Only expose `@Prop` annotated fields for Spec components
        continue;
      }

      hasProps = true;

      if (prop != null && PropWithInspectorSection.class.isAssignableFrom(prop.getClass())) {
        final AbstractMap.SimpleEntry<String, String> datum =
            ((PropWithInspectorSection) prop).getFlipperLayoutInspectorSection();
        if (datum != null) {
          data.add(new Named<>(datum.getKey(), new FlipperObject(datum.getValue())));
        }
      }

      if (annotation != null) {
        ResType resType = annotation.resType();
        if (resType == ResType.COLOR) {
          props.put(propName, prop == null ? "null" : fromColor((Integer) prop));
          continue;
        } else if (resType == ResType.DRAWABLE) {
          props.put(propName, prop == null ? "null" : fromDrawable((Drawable) prop));
          continue;
        }
      }

      if (prop != null && PropWithDescription.class.isAssignableFrom(prop.getClass())) {
        final Object description =
            ((PropWithDescription) prop).getFlipperLayoutInspectorPropDescription();
        // Treat the description as immutable for now, because it's a "translation" of the
        // actual prop,
        // mutating them is not going to change the original prop.
        if (description instanceof Map<?, ?>) {
          final Map<?, ?> descriptionMap = (Map<?, ?>) description;
          for (Map.Entry<?, ?> entry : descriptionMap.entrySet()) {
            props.put(entry.getKey().toString(), InspectorValue.immutable(entry.getValue()));
          }
        } else {
          props.put(propName, InspectorValue.immutable(description));
        }
        continue;
      }

      props.put(propName, FlipperEditor.makeFlipperField(node, f));
    }

    if (hasProps) {
      data.add(new Named<>("Props", props.build()));
    }

    return data;
  }

  @Nullable
  static FlipperObject getStateData(StateContainer stateContainer) {
    if (stateContainer == null) {
      return null;
    }

    final FlipperObject.Builder state = new FlipperObject.Builder();

    boolean hasState = false;
    for (Field f : stateContainer.getClass().getDeclaredFields()) {
      f.setAccessible(true);

      final State annotation = f.getAnnotation(State.class);
      if (annotation != null) {
        state.put(f.getName(), FlipperEditor.makeFlipperField(stateContainer, f));
        hasState = true;
      }
    }

    return hasState ? state.build() : null;
  }

  static InspectorValue fromDrawable(Drawable d) {
    int color = 0;
    if (d instanceof ColorDrawable) {
      color = ((ColorDrawable) d).getColor();
    } else if (d instanceof ComparableColorDrawable) {
      color = ((ComparableColorDrawable) d).getColor();
    }
    return InspectorValue.mutable(Color, color);
  }

  static InspectorValue fromColor(int color) {
    return InspectorValue.mutable(Color, color);
  }
}

File: litho-editor-flipper/src/main/java/com/facebook/litho/editor/flipper/DataUtils.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho.editor.flipper

import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import com.facebook.flipper.core.FlipperObject
import com.facebook.flipper.plugins.inspector.InspectorValue
import com.facebook.flipper.plugins.inspector.InspectorValue.Type.Color
import com.facebook.flipper.plugins.inspector.Named
import com.facebook.litho.Component
import com.facebook.litho.DebugComponent
import com.facebook.litho.KStateContainer
import com.facebook.litho.SpecGeneratedComponent
import com.facebook.litho.StateContainer
import com.facebook.litho.annotations.Prop
import com.facebook.litho.annotations.ResType
import com.facebook.litho.annotations.State
import com.facebook.litho.state

object DataUtils {

  @JvmStatic
  fun getPropData(node: Any): List<Named<FlipperObject>> {
    val props = FlipperObject.Builder()
    val data: MutableList<Named<FlipperObject>> = ArrayList()
    var hasProps = false
    val isSpecComponent = node is SpecGeneratedComponent
    for (f in node.javaClass.declaredFields) {
      f.isAccessible = true
      val prop =
          try {
            f[node]
          } catch (e: IllegalAccessException) {
            continue
          }
      val propName = f.name
      val annotation = f.getAnnotation(Prop::class.java)
      if (isSpecComponent && annotation == null) {
        // Only expose `@Prop` annotated fields for Spec components
        continue
      }
      hasProps = true
      if (prop != null && PropWithInspectorSection::class.java.isAssignableFrom(prop.javaClass)) {
        val datum = (prop as PropWithInspectorSection).flipperLayoutInspectorSection
        if (datum != null) {
          data.add(Named(datum.key, FlipperObject(datum.value)))
        }
      }
      if (annotation != null) {
        val resType = annotation.resType
        if (resType == ResType.COLOR) {
          props.put(propName, if (prop == null) "null" else fromColor(prop as Int))
          continue
        } else if (resType == ResType.DRAWABLE) {
          props.put(propName, if (prop == null) "null" else fromDrawable(prop as Drawable?))
          continue
        }
      }

      val description = getValueOverride(prop)
      if (description != null) {
        props.put(propName, description)
        continue
      }

      props.put(propName, FlipperEditor.makeFlipperField(node, f))
    }
    if (hasProps) {
      data.add(Named("Props", props.build()))
    }
    return data
  }

  @JvmStatic
  fun getMountingData(node: DebugComponent): FlipperObject? {

    val lithoView = node.lithoView
    val mountingData = FlipperObject.Builder()

    if (lithoView == null) {
      return mountingData.build()
    }

    val mountState = lithoView.mountDelegateTarget ?: return mountingData.build()
    val componentTree = lithoView.componentTree ?: return mountingData.build()
    var hasMountingInfo = false
    val component = node.component

    if (component.mountType != Component.MountType.NONE) {
      val renderUnit = DebugComponent.getRenderUnit(node, componentTree)
      if (renderUnit != null) {
        val renderUnitId = renderUnit.id
        val isMounted = mountState.getContentById(renderUnitId) != null
        mountingData.put("mounted", isMounted)
        hasMountingInfo = true
      }
    }

    val visibilityOutput = DebugComponent.getVisibilityOutput(node, componentTree)
    if (visibilityOutput != null) {
      val isVisible = DebugComponent.isVisible(node, lithoView)
      mountingData.put("visible", isVisible)
      hasMountingInfo = true
    }

    return if (hasMountingInfo) mountingData.build() else null
  }

  // Props can override the value shown in the Layout Inspector by implementing PropWithDescription
  // e.g. to include additional debug information.
  private fun getValueOverride(prop: Any?): FlipperObject? =
      prop
          ?.let { it as? PropWithDescription }
          ?.flipperLayoutInspectorPropDescription
          ?.let {
            when (it) {
              is Map<*, *> ->
                  FlipperObject.Builder()
                      .apply {
                        it.entries.forEach { (key, value) ->
                          put(key.toString(), InspectorValue.immutable(value))
                        }
                      }
                      .build()
              else -> InspectorValue.immutable(it).toFlipperObject()
            }
          }

  @JvmStatic
  fun getStateData(stateContainer: StateContainer?): FlipperObject? {
    stateContainer ?: return null

    val state = FlipperObject.Builder()
    var hasState = false
    if (stateContainer is KStateContainer) {
      val states = stateContainer.state
      hasState = states.isNotEmpty() == true
      states?.let {
        it.forEachIndexed { index, item ->
          // TODO(T131825073): Replace with Flipper Field to allow editing state
          state.put(index.toString(), item)
        }
      }
    } else {
      for (f in stateContainer.javaClass.declaredFields) {
        f.isAccessible = true
        val annotation = f.getAnnotation(State::class.java)
        if (annotation != null) {
          state.put(f.name, FlipperEditor.makeFlipperField(stateContainer, f))
          hasState = true
        }
      }
    }

    return if (hasState) state.build() else null
  }

  @JvmStatic
  fun fromDrawable(d: Drawable?): InspectorValue<*> =
      when (d) {
        is ColorDrawable -> InspectorValue.immutable(Color, d.color)
        else -> InspectorValue.immutable(d)
      }

  @JvmStatic fun fromColor(color: Int): InspectorValue<*> = InspectorValue.mutable(Color, color)
}
